handlers/mist_triggers.go

```diff
@@ -0,0 +1,207 @@
+		}
+	}
+}
+
+// TriggerPushEnd responds to PUSH_END trigger
+// This trigger is run whenever an outgoing push stops, for any reason.
+// This trigger is stream-specific and non-blocking. The payload for this trigger is multiple lines,
+// each separated by a single newline character (without an ending newline), containing data:
+//   push ID (integer)
+//   stream name (string)
+//   target URI, before variables/triggers affected it (string)
+//   target URI, afterwards, as actually used (string)
+//   last 10 log messages (JSON array string)
+//   most recent push status (JSON object string)
+func (d *MistCallbackHandlersCollection) TriggerPushEnd(t *Trigger) {
+	payload, err := io.ReadAll(t.req.Body)
+	if err != nil {
+		t.Err("Cannot read payload", err)
+		return
+	}
+	lines := strings.Split(strings.TrimSuffix(string(payload), "\n"), "\n")
+	if len(lines) != 6 {
+		errors.WriteHTTPBadRequest(t.w, "Bad request payload", fmt.Errorf("unknown payload '%s'", string(payload)))
+		return
+	}
+	// stream name is the second line in the Mist Trigger payload
+	t.streamName = lines[1]
+	destination := lines[2]
+	actualDestination := lines[3]
+	pushStatus := lines[5]
+	if yes, _ := isTranscodeStream(t.streamName); yes {
+		d.TranscodingPushEnd(t, destination, actualDestination, pushStatus)
+	} else {
+		d.SegmentingPushEnd(t)
+	}
+}
+
+func (d *MistCallbackHandlersCollection) TranscodingPushEnd(t *Trigger, destination, actualDestination, pushStatus string) {
+	info, err := d.StreamCache.Transcoding.Get(t.streamName)
+	if err != nil {
+		t.Err("unknown push source", err)
+		return
+	}
+	inOurBooks := false
+	for i := 0; i < len(info.Destionations); i++ {
+		if info.Destionations[i] == destination {
+			inOurBooks = true
+			break
+		}
+	}
+	if !inOurBooks {
+		t.Err("missing from our books", nil)
+		return
+	}
+	callbackClient := clients.NewCallbackClient()
+	if uploadSuccess := pushStatus == "null"; uploadSuccess {
+		if err := callbackClient.SendRenditionUpload(info.CallbackUrl, info.Source, actualDestination); err != nil {
+			t.Err("Cannot send rendition transcode status", err)
+		}
+	} else {
+		// We forward pushStatus json to callback
+		if err := callbackClient.SendRenditionUploadError(info.CallbackUrl, info.Source, actualDestination, pushStatus); err != nil {
+			t.Err("Cannot send rendition transcode error", err)
+		}
+	}
+	// We do not delete triggers as source stream is wildcard stream: RENDITION_PREFIX
+	if empty := d.StreamCache.Transcoding.RemovePushDestination(t.streamName, destination); empty {
+		if err := callbackClient.SendSegmentTranscodeStatus(info.CallbackUrl, info.Source); err != nil {
+			t.Err("Cannot send segment transcode status", err)
+		}
+		d.StreamCache.Transcoding.Remove(t.streamName)
+	}
+	return
+}
+
+func (d *MistCallbackHandlersCollection) SegmentingPushEnd(t *Trigger) {
+	// when uploading is done, remove trigger and stream from Mist
+	errT := d.MistClient.DeleteTrigger(t.streamName, "PUSH_END")
```

Comment:
I think it'd safer to first notify callbackClient and then delete triggers / streams for Mist. If for some reason the deletion fails, then I think the callback client should still get informed about the process.

Reply:
Makes sense.

Note i wasn't changing segmenting logic in this PR, just mechanisms around it for trigger callbacks, error reporting and similar..